### PR TITLE
Add Chunked ORPO Loss

### PR DESCRIPTION
**Pull request by shivam15s**
## Summary

Adds chunked ORPO loss kernel 

## Testing Done

Benchmarks
Speed ORPO
Mem ORPO

References:
1. #227 
2. https://gist.github.com/Chillee/22cd93e11b887db1f596ab754d60a899#file-lce_benchmark-py

- Hardware Type: <BLANK>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    )
+    ratio = F.logsigmoid(log_odds)
+    return beta * ratio.sum()
+
+
+class LigerFusedLinearORPOFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, _input, weight, target, bias=None, ignore_index=-100, beta=0.1, compiled=True):
+        """
+        Fused linear forward function with ORPO (Odds-Ratio Preference Optimization).
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:31]: Can you please add a reference link to the paper or a blog explaining the function.

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        Args:
+            _input (torch.Tensor): Input tensor. Shape: (batch_size, seq_len, hidden_size).
+            weight (torch.Tensor): Weight tensor. Shape: (vocab_size, hidden_size).
+            bias (torch.Tensor, optional): Bias tensor. Shape: (hidden_size,).
+            ignore_index (int): Index to ignore for loss computation.
+            compiled (bool): Whether to use compiled mode for chunk accumulation.
+        """
+        CHUNK_SIZE = 1
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:39]: Should we add a TODO to tune this param? We might be under-utilizing GPU memory if we number of chunks is always set to batch_size/2.

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+class LigerFusedLinearORPOFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, _input, weight, target, bias=None, ignore_index=-100, beta=0.1, compiled=True):
+        """
+        Fused linear forward function with ORPO (Odds-Ratio Preference Optimization).
+        Args:
+            _input (torch.Tensor): Input tensor. Shape: (batch_size, seq_len, hidden_size).
+            weight (torch.Tensor): Weight tensor. Shape: (vocab_size, hidden_size).
+            bias (torch.Tensor, optional): Bias tensor. Shape: (hidden_size,).
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:35]: Shouldn't shape of the bias term be = `vocab_size` ?

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            unnorm_logits = input_chunk @ weight.t()  # chunk_size x V
+            if bias is not None:
+                unnorm_logits = unnorm_logits + bias
+            unnorm_logits = unnorm_logits.float()
+            norm_logits = F.log_softmax(unnorm_logits, dim=-1)
+
+            chosen_nll_loss = F.nll_loss(
+                norm_logits[:len_chosen_chunk].view(-1, norm_logits.shape[-1]),
+                target_chunk[:len_chosen_chunk].view(-1),
+                reduction="sum",
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:53]: wondering if this should be a customizable choice between `mean` and `sum`

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                ignore_index=ignore_index,
+            )
+            chosen_nll_loss = chosen_nll_loss / (target[:target.shape[0]//2] != ignore_index).sum()
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:56]: If my understanding is correct, this is computing the mean nll loss for the chosen/accepted response, why can't we just use 'mean' as the reduction method above rather than computing the sum first and then dividing by the number of non-pad tokens?

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 3 above)
Comment (pramodith) [reply src/liger_kernel/chunked_loss/orpo_loss.py:35]: We're also missing a doc string for `target`

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        grad_chosen_inputs = []
+        grad_rejected_inputs = []
+        grad_bias = torch.zeros_like(bias) if bias is not None else None
+        loss_acc = torch.zeros((), device=_input.device)
+
+        chunks = max(1, _input.shape[0] // (2 * CHUNK_SIZE))
+
+        def accumulate_chunk(input_chunk, target_chunk):
+            if bias is not None:
+                (chunk_grad_input, chunk_grad_weight, chunk_grad_bias), chunk_loss = torch.func.grad_and_value(
```
Comment (pramodith) [inline src/liger_kernel/chunked_loss/orpo_loss.py:86]: TIL about `torch.func.grad_and_value`! Cool stuff!

File: test/chunked_loss/test_orpo_loss.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        labels = torch.where(labels == self.ignore_index, 0, labels)
+
+        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)).squeeze(2)
+
+        if average_log_prob:
+            return (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
+        else:
+            return (per_token_logps * loss_mask).sum(-1)
+
+    def odds_ratio_loss(
```
Comment (pramodith) [inline test/chunked_loss/test_orpo_loss.py:46]: We should probably add a link to the original source of the code: https://github.com/huggingface/trl/blob/v0.11.3/trl/trainer/orpo_trainer.py#L75

File: test/chunked_loss/test_orpo_loss.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    [
+        (1.0, torch.bfloat16, 5e-2, 5e-1),
+        (1.0, torch.float32, 1e-5, 5e-4),
+    ],
+)
+@pytest.mark.parametrize("bias", [True, False])
+@pytest.mark.parametrize("ignore_index, beta", [(-100, 0.1), (42, 0.2)])
+def test_correctness(B, T, H, V, scalar, dtype, atol, rtol, bias, ignore_index, beta):
+    # Define input tensors
+    _tensor = torch.randn(B, T, H, device="cuda", dtype=dtype) * scalar
```
Comment (pramodith) [inline test/chunked_loss/test_orpo_loss.py:163]: What's the purpose of scalar?

**Review by pramodith — CHANGES_REQUESTED** [review]
Great stuff! Most of my comments are around documentation, wondering if we should also include a benchmark script for this one!

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 5 above)
Comment (shivam15s) [reply src/liger_kernel/chunked_loss/orpo_loss.py:56]: The target variable here is the whole target while target_chunk is the target for a given chunk. Normalizing directly using nll_loss mean reduction would normalize with target_chunk length which is mathematically wrong. So this is kinda hacky way to do normalization

File: test/chunked_loss/test_orpo_loss.py (hunk 8 above)
Comment (shivam15s) [reply test/chunked_loss/test_orpo_loss.py:163]: I believe the intention is to support testing inputs with variances other than one. However, the current tests do not yet include cases with non-unit variances.

**Comment by lancerts** [comment]
> Great stuff! Most of my comments are around documentation, wondering if we should also include a benchmark script for this one!

@shivam15s great work! @pramodith thanks for the detailed review. 
And yes, we should add the bench script

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 5 above)
Comment (pramodith) [reply src/liger_kernel/chunked_loss/orpo_loss.py]: Ahh gotcha, that makes sense, thanks for the explanation.

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+            per_token_logps = norm_logits.gather(-1, label_chunk.unsqueeze(-1)).squeeze(
+                -1
+            )
+            average_log_prob = (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
+
+            chosen_logps = average_log_prob[:len_chosen_chunk]
+            rejected_logps = average_log_prob[len_chosen_chunk:]
+
+            or_loss = odds_ratio_loss(chosen_logps, rejected_logps, beta=beta)
```
Comment (Xiaoling1992) [inline src/liger_kernel/chunked_loss/orpo_loss.py:86]: could we rename this to odds ratio loss?

The returned is orpo loss = sft loss + odds ratio loss, not this one.

File: src/liger_kernel/chunked_loss/orpo_loss.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+            or_loss = or_loss / (target.shape[0] // 2)
+
+            loss = chosen_nll_loss - or_loss
```
Comment (Xiaoling1992) [inline src/liger_kernel/chunked_loss/orpo_loss.py:89]: Could this class take a flag like enable_sft_loss? 

Since users might have their customized sft loss logic and they needs odds ratio loss only. When the flag is True, this class computes chosen_nll_loss and returns chosen_nll_loss - or_loss, otherwise, it computes and returns odds_ratio_loss only.

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 9 above)
Comment (shivam15s) [reply src/liger_kernel/chunked_loss/orpo_loss.py:86]: To clarify, or_loss here is the odds ratio loss; while the loss returned from the fn is the sum of odds ratio loss (or_loss) and sft loss (chosen_nll_loss). Do you want to rename or_loss to something more verbose?

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 10 above)
Comment (lancerts) [reply src/liger_kernel/chunked_loss/orpo_loss.py:89]: +1, or we can add an input parameter `chosen_nll_loss_weight` and `loss = chosen_nll_loss_weight* chosen_nll_loss - or_loss`

Comment (Xiaoling1992) [reply src/liger_kernel/chunked_loss/orpo_loss.py]: There is already a beta to balance the weights between sft loss and odds_ratio loss.
loss = sft loss + beta * odds ratio loss

File: src/liger_kernel/chunked_loss/orpo_loss.py (hunk 9 above)
Comment (Xiaoling1992) [reply src/liger_kernel/chunked_loss/orpo_loss.py]: Thanks for clarification. Could we rename or_loss to odds_ratio_loss since or_loss might be misunderstood to be orpo_loss.

File: src/liger_kernel/chunked_loss/fused_linear_preference.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
+import torch
+
+
+class LigerFusedLinearPreferenceBase(torch.autograd.Function):
+    @staticmethod
+    def forward(
```
Comment (ByronHsu) [inline src/liger_kernel/chunked_loss/fused_linear_preference.py:6]: general suggestion:

1. can we make loss_fn a member method in the base class and raise not implemented exception and the derive class needs to implement it? instead of being a separate python function

2. can we use `__super__` instead of `LigerFusedLinearPreferenceBase.forward` (backward). Or we can not do that for static method.


I will merge this PR first and we can address later! Great contribution!!

**Review by ByronHsu — APPROVED** [review]
